### ISSUE TITLE
StringMatchingRecommender does not notice when it has no training data

### ISSUE THREAD

**Issue by reckart** [inception-project/inception#853]
**Describe the bug**
The `StringMatchingRecommender` currently generates a sample for each sentence it encounters, no matter if that sentence actually contains annotations. This means that the check for the training/test set sizes will pass even if there is no training data at all.

**To Reproduce**
1. Import a text
2. Setup a StringMatchingRecommender to predict named entities
3. Open document and watch evaluation

**Expected behavior**
If there is no training data or too little training data, the recommender should log a corresponding message and return `0.0`.

**Please complete the following information:**
 - Version and build ID: a6f2bd5a795f0120922114b88d85d89ab96f4b9d

**Additional context**
When changing the code such that it only considers sentences with at least one annotation, the evaluation results change since samples are not only used for training but also for evaluation. Considering only sentences with annotations as samples leaves out all gold-standard sentences which have no annotations at all meaning that the recommender cannot make false-positive predictions on them. 

**Running `thatIncrementalNerEvaluationWorks` only considering sentences with at least one annotation:**
```
2019-01-12 21:01:09 TRACE StringMatchingRecommender - Extracting data took 488ms
2019-01-12 21:01:09 INFO StringMatchingRecommender - Training on [5000] items, predicting on [3687] of total [18421]
Score: 0,260814
2019-01-12 21:01:12 TRACE StringMatchingRecommender - Extracting data took 323ms
2019-01-12 21:01:12 INFO StringMatchingRecommender - Training on [10000] items, predicting on [3687] of total [18421]
Score: 0,317409
2019-01-12 21:01:14 TRACE StringMatchingRecommender - Extracting data took 244ms
2019-01-12 21:01:14 INFO StringMatchingRecommender - Training on [14734] items, predicting on [3687] of total [18421]
Score: 0,347855
```

**Running `thatIncrementalNerEvaluationWorks` only considering all sentences:**
```
2019-01-12 21:09:34 TRACE StringMatchingRecommender - Extracting data took 510ms
2019-01-12 21:09:34 INFO StringMatchingRecommender - Training on [5000] items, predicting on [6263] of total [31300]
Score: 0,219644
2019-01-12 21:09:37 TRACE StringMatchingRecommender - Extracting data took 283ms
2019-01-12 21:09:37 INFO StringMatchingRecommender - Training on [10000] items, predicting on [6263] of total [31300]
Score: 0,270006
2019-01-12 21:09:39 TRACE StringMatchingRecommender - Extracting data took 265ms
2019-01-12 21:09:39 INFO StringMatchingRecommender - Training on [15000] items, predicting on [6263] of total [31300]
Score: 0,309062
```

The `OpenNlpPosRecommender` currently considers any sentence with at least one POS annotation as a sample. So does the `OpenNlpNerRecommender`.

**The question is**:
- should we generate samples for all sentences - even those without annotations - but when checking the training set for its size consider only those samples which contain annotations?
- should we generate samples for all sentences - but then drop the unannotated samples from the training set so that learner doesn't even see them?
- ...?

**Comment by Rentier** [comment]
I think it depends on the kind of metric you want to use. If you want to just check whether the gold labels are predicted, then we can use accuracy and just compare whether for the gold labels, the predictions are correct. If you want to compute agreement, then we need unannotated sentences also in the evaluation set. I fear that any dropping after the split leads to a different split than wished. I would do both of your options then, the second implies a bit the first imho
